ultravox/tools/ds_tool/ds_tool.py

```diff
@@ -142,32 +180,40 @@ def main(args: DatasetToolArgs):
     data_dict: datasets.DatasetDict = datasets.load_dataset(
         ds_name, args.dataset_subset, split=args.dataset_split
     )
-    if args.dataset_split:
-        data_dict = datasets.DatasetDict(**{args.dataset_split: data_dict})
+
+    if isinstance(data_dict, datasets.Dataset):
+        data_dict = datasets.DatasetDict({args.upload_split: data_dict})
+
+    if len(data_dict) > 1 and args.upload_split:
+        raise ValueError("Cannot upload multiple splits to a single split")
 
     for split, ds_split in data_dict.items():
-        print(f'Processing split "{split}"...')
+        print(
+            f"Processing dataset: {ds_name}, subset {args.dataset_subset}, split {args.dataset_split}, containing {len(ds_split)} samples"
+        )
         if args.shuffle:
             ds_split = ds_split.shuffle(seed=args.shuffle_seed)
         if args.num_samples:
             ds_split = ds_split.select(range(args.num_samples))
-        data_dict[split] = args.task.map_split(ds_split, args.num_workers)
-
-    token = args.token or os.environ.get("HF_TOKEN")
+        data_dict[split] = args.task.map_split(
+            ds_split, args.num_workers, args.writer_batch_size
+        )
+    
     hub_args: Dict[str, Any] = {
-        "config_name": args.dataset_subset or "default",
-        "token": token,
+        "config_name": args.upload_subset if args.upload_subset else "default",
+        "token": args.token or os.environ.get("HF_TOKEN"),
         "revision": args.upload_branch,
         "private": args.private,
-        "split": args.upload_split,
     }
-    if args.num_shards is not None:
-        hub_args["num_shards"] = {split: args.num_shards for split in data_dict.keys()}
 
+    if args.num_shards is not None:
+        hub_args["num_shards"] = args.num_shards
+    
     try:
         if args.dataset_split:
+            upload_split = args.upload_split or args.dataset_split
```

Comment:
I think `upload_split` will already be set from `__post_init__`, right?

Reply:
good catch